source/fuzz/fuzzer_pass_add_image_sample_unused_components.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2020 André Perez Maselco
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/fuzzer_pass_add_image_sample_unused_components.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+#include "source/fuzz/transformation_add_image_sample_unused_components.h"
+#include "source/fuzz/transformation_composite_construct.h"
+
+namespace spvtools {
+namespace fuzz {
+
+FuzzerPassAddImageSampleUnusedComponents::
+    FuzzerPassAddImageSampleUnusedComponents(
+        opt::IRContext* ir_context,
+        TransformationContext* transformation_context,
+        FuzzerContext* fuzzer_context,
+        protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, transformation_context, fuzzer_context,
+                 transformations) {}
+
+FuzzerPassAddImageSampleUnusedComponents::
+    ~FuzzerPassAddImageSampleUnusedComponents() = default;
+
+void FuzzerPassAddImageSampleUnusedComponents::Apply() {
+  GetIRContext()->module()->ForEachInst([this](opt::Instruction* instruction) {
+    if (!spvOpcodeIsImageSample(instruction->opcode())) {
+      return;
+    }
+
+    if (!GetFuzzerContext()->ChoosePercentage(
+            GetFuzzerContext()
+                ->GetChanceOfAddingImageSampleUnusedComponents())) {
+      return;
+    }
+
```

Comment:
The following looks very good.  However, it is intricate and will be hard to understand for someone who doesn't already know this instruction.

I suggest you add two worked examples, one where the existing type is int, the other where it is vec2 (i.e., a 2d floating point vector).  Then at each stage of the algorithm you show using psesudocode what would be computed in each case.

Take a look at what I've done in transformation_add_function.cpp from lines 586-614 - there are little snippets of illustrative SPIR-V.  I think something along those lines could be useful here.  It will be different as you'll be describing existing SPIR-V and the types that will be needed based on it.

Reply:
Don't you think the comments explain by themselves?
I'm trying to explain with code, but I guess it is redundant.

Reply:
For me, a worked example alongside the comments would be really useful.  The comments explain what the code does *in general*.   A worked example would show what the code does to a specific fragment of SPIR-V.  I think the two together would be great.

Can you have a go?

Reply:
Done.